sig-architecture/generics.md

```diff
@@ -0,0 +1,29 @@
+# Background
+
+In general, Kubernetes has not restricted using new Go features, we've quickly
+adopted new standard library types and methods and will continue to do so.
+
+Generally the latest stable go release is in use on the main development branch.
+This includes all of the staging libraries (client-go etc.) that originate in the
+main [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes) repository.
+
+In Kubernetes v1.24 we shipped Go 1.18 (which adds support for generics)
+rather late in the release cycle, so we temporarily prohibited using generics
+in case we ran into other issues and needed to roll back to unblock the release.
+Now that v1.24.0 is out, use of generics should be unrestricted.
+
+# Generics Policy
```

Comment:
Great point, we only discussed this from the staging library angle previously.

We should add a note similar to the bug fixes guidance, but as a hard requirement, I think.

Reply:
I rewrote this a bit to position limitations until v1.24 is out of support and address this point.

Reply:
looks fine as a starting point. we could maybe relax the guidance to allow use of generics in those repos in ways that would still supports go1.17, but I'm not familiar with how possible/fiddly that is... we'd want to be really confident in the implications before doing so, I guess

Reply:
Even if technically possible (I think so with build tags?) I don’t think we want to allow this versus waiting a few more releases just because it’s very similar to starting to support multiple branches of these libraries from a support perspective and we don’t do that today in any I can think of.